src/components/shared/Validations.tsx

```diff
@@ -1,90 +1,161 @@
-      {visibility}
-      {validationCount ? (
-        <Popover
-          boundary="window"
-          interactionKind="hover"
-          content={
-            <div className="p-5" style={{ maxHeight: 500, maxWidth: 400 }}>
-              {Object.keys(validations).map((key, index) => {
-                const validation = validations[key];
+      <NumberValidations validations={numberValidations} />
+      <KeyValueValidations validations={keyValueValidations} />
+      <NameValidations validations={booleanValidations} />
+    </>
+  );
+};
 
-                let elem = null;
-                if (Array.isArray(validation)) {
-                  elem = validation.map((v, i) => (
-                    <div key={i} className="mt-1 mr-1 flex items-center">
-                      <div className="px-1 bg-gray-2 dark:bg-gray-8 font-bold text-sm rounded">{String(v)}</div>
-                      {i < validation.length - 1 ? <div>,</div> : null}
-                    </div>
-                  ));
-                } else if (typeof validation === 'object') {
-                  elem = (
-                    <div className="m-1 px-1 bg-gray-2 dark:bg-gray-8 font-bold text-sm rounded" key={index}>
-                      {'{...}'}
-                    </div>
-                  );
-                } else {
-                  elem = (
-                    <div className="m-1 px-1 bg-gray-2 dark:bg-gray-8 font-bold text-sm rounded" key={index}>
-                      {typeof validation === 'string' ? `"${validation}"` : safeStringify(validation)}
-                    </div>
-                  );
-                }
+const NumberValidations = ({ validations, className }: { validations: Dictionary<unknown>; className?: string }) => {
+  const values = keys(omit(validations, excludedValidations));
+  if (!values.length) {
+    return null;
+  }
+  return (
+    <Flex my={2} color="muted">
+      <Text fontWeight="light">Allowed values:</Text>
+      {values.map(key => {
+        let suffix;
+        if (key.includes('Length')) {
+          suffix = ' characters';
+        } else if (key.includes('Items')) {
+          suffix = ' items';
+        } else {
+          suffix = '';
+        }
 
-                return (
-                  <div key={index} className="py-1 flex items-baseline">
-                    <div className="font-medium pr-2">{key}:</div>
-                    <div className="flex-1 flex flex-wrap justify-end">{elem}</div>
-                  </div>
-                );
-              })}
-            </div>
-          }
-          target={requiredElem}
-        />
-      ) : (
-        requiredElem
-      )}
-    </>
+        const exclusive =
+          (key === 'minimum' && validations.exclusiveMinimum) || (key === 'maximum' && validations.exclusiveMaximum)
+            ? true
+            : false;
+        const sign = `${key.includes('min') ? '>' : '<'}${exclusive ? '' : '='}`;
```

Comment:
This one is completely optional but there's a bit too much magic going on to my taste here, and some of the magic depends on property names like the definition of `sign`, which might just be a bit too brittle for future maintenance.

As there are only 8 number validation formats defined, maybe defining them one-by-one won't be much longer but more readable.

Something like:
```tsx
// this typing requires `numberValidationNames` to be defined `as const`.
const numberValidationFormatters: Record<typeof numberValidationNames[number], (value: undefined) => string> = {
  minimum: (value) => `>= ${value}`,
  exclusiveMinimum: (value) => `> ${value}`,
  minItems:(value) => `>= ${value} items`,
  minLength:(value) => `>= ${value} characters`,
  maximum: (value) => `<= ${value}`,
  exclusiveMaximum: (value) => `< ${value}`,
  maxItems: (value) => `< ${value} items`,
  maxLength: (value) => `< ${value} characters`,
}

const NumberValidations = ({ validations, className }: { validations: Partial<Record<typeof numberValidationNames[number], unknown>>; className?: string }) => {
  const entries = Object.entries(validations);
  if (!entries.length) {
    return null;
  }
  return (
    <Flex my={2} color="muted">
      <Text fontWeight="light">Allowed values:</Text>
      {entries.map(([key, value]) => numberValidationFormatters[key](value))
        .map((value, i) => (<Text key={i} ml={2} px={1} fontFamily="mono" border rounded="lg" className={className}>{value}</Text>))}
    </Flex>
  )
};
```

This makes it much easier to add new validations or change the copy of existing ones.

Reply:
I like your approach. That is definitely more readable. Already updated